synthesizer/process/src/cost.rs

```diff
@@ -404,60 +407,60 @@ pub fn cost_per_command<N: Network>(
-    // Aggregate the cost of all commands in the program.
-    finalize
-        .commands()
-        .iter()
-        .map(|command| cost_per_command(stack, finalize, command, ConsensusFeeVersion::V2))
-        .try_fold(future_cost, |acc, res| {
-            res.and_then(|x| acc.checked_add(x).ok_or(anyhow!("Finalize cost overflowed")))
-        })
+    cost_in_microcredits(stack, function_name, ConsensusFeeVersion::V2)
 }
 
-/// Returns the minimum number of microcredits required to run the finalize (depcrated).
+/// Returns the minimum number of microcredits required to run the finalize (deprecated).
 pub fn cost_in_microcredits_v1<N: Network>(stack: &Stack<N>, function_name: &Identifier<N>) -> Result<u64> {
-    // Retrieve the finalize logic.
-    let Some(finalize) = stack.get_function_ref(function_name)?.finalize_logic() else {
-        // Return a finalize cost of 0, if the function does not have a finalize scope.
-        return Ok(0);
-    };
-    // Get the cost of finalizing all futures.
-    let mut future_cost = 0u64;
-    for input in finalize.inputs() {
-        if let FinalizeType::Future(future) = input.finalize_type() {
-            // Get the external stack for the future.
-            let stack = stack.get_external_stack(future.program_id())?;
-            // Accumulate the finalize cost of the future.
-            future_cost = future_cost
-                .checked_add(cost_in_microcredits_v1(&stack, future.resource())?)
-                .ok_or(anyhow!("Finalize cost overflowed"))?;
+    cost_in_microcredits(stack, function_name, ConsensusFeeVersion::V1)
+}
+
+// A helper function to compute the cost in microcredits for a given function.
+fn cost_in_microcredits<N: Network>(
+    stack: &Stack<N>,
+    function_name: &Identifier<N>,
+    consensus_fee_version: ConsensusFeeVersion,
+) -> Result<u64> {
+    // Initialize the base cost.
+    let mut finalize_cost = 0u64;
+    // Initialize a queue of finalize blocks to tally.
+    let mut finalizes = vec![(StackRef::Internal(stack), *function_name)];
+    // Initialize a counter for the number of finalize blocks seen.
+    let mut num_finalizes = 1;
+    // Iterate over the finalize blocks.
+    while let Some((stack_ref, function_name)) = finalizes.pop() {
+        // Ensure that the number of finalize blocks does not exceed the maximum.
+        // Note that one transition is reserved for the fee.
+        ensure!(
+            num_finalizes < Transaction::<N>::MAX_TRANSITIONS,
+            "The number of finalize blocks must be less than '{}'",
+            Transaction::<N>::MAX_TRANSITIONS
+        );
+        // Get the finalize logic. If the function does not have a finalize scope then no cost is incurred.
+        if let Some(finalize) = match &stack_ref {
+            StackRef::Internal(stack_ref) => stack_ref.get_function_ref(&function_name)?.finalize_logic(),
+            StackRef::External(stack_ref) => stack_ref.get_function_ref(&function_name)?.finalize_logic(),
```

Comment:
it would be best if we could avoid separate `match` branches here, as their logic is the same